[JSC] Extend JSBigInt::compare and JSBigInt::compareToDouble for Int64 and Int32

hyjorc1 opened the pull request:
#### f706066e1e24e20a84f82374480a7b33e1de2d10
<pre>
[JSC] Extend JSBigInt::compare and JSBigInt::compareToDouble for Int64 and Int32
<a href="https://bugs.webkit.org/show_bug.cgi?id=271073">https://bugs.webkit.org/show_bug.cgi?id=271073</a>
<a href="https://rdar.apple.com/124706111">rdar://124706111</a>

Reviewed by Keith Miller.

This patch extends JSBigInt::compare and JSBigInt::compareToDouble
for Int64 and Int32.

* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::Int64BigIntImpl::Int64BigIntImpl):
(JSC::Int64BigIntImpl::isZero):
(JSC::Int64BigIntImpl::sign):
(JSC::Int64BigIntImpl::length):
(JSC::Int64BigIntImpl::digit):
(JSC::JSBigInt::compare):
(JSC::JSBigInt::compareToDouble):
* Source/JavaScriptCore/runtime/JSBigInt.h:

Canonical link: <a href="https://commits.webkit.org/276417@main">https://commits.webkit.org/276417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8ad48fda6671634adc01e4dd689f8935e9e2a345

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6152 "Built successfully and passed tests") | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -473,6 +473,39 @@ class Int32BigIntImpl {
+
+    ALWAYS_INLINE bool isZero() { return !m_value; }
+    ALWAYS_INLINE bool sign() { return m_sign; }
+    ALWAYS_INLINE unsigned length() { return !isZero() ? 0 : 1; }
```
It should be
```
ALWAYS_INLINE unsigned length() { return isZero() ? 0 : 1; }
```

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -473,6 +473,39 @@ class Int32BigIntImpl {
+
+private:
+    friend struct JSBigInt::ImplResult;
+    JSBigInt::Digit m_value;
```
Isn't this too small for 32-bit platforms since `JSBigInt::Digit = UCPURegister`?

We should probably just use uint64_t and fix `length()` and `digit`

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -1439,15 +1472,68 @@ JSBigInt::ComparisonResult JSBigInt::compare(JSBigInt* x, JSBigInt* y)
+#if USE(BIGINT32)
+    if (x.isBigInt32() && y.isBigInt32()) {
+        int32_t x1 = x.isBigInt32();
+        int32_t y1 = y.isBigInt32();
```
This should be `x.asBigInt32()`.

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -473,6 +475,55 @@ class Int32BigIntImpl {
+    friend struct JSBigInt::ImplResult;
+    uint64_t m_value;
+    bool m_sign;
+};
```
I'd do 
```c++
// In WTF/MathExtras.h:

// For use in places where we could negate std::numeric_limits<T>::min and would like to avoid UB.
template<std::integral T>
ALWAYS_INLINE T negate(T v)
{
    return ~static_cast<std::make_unsigned_t<T>>(v) + 1;
}

// Here:

class Int64BigIntImpl {
public:
    explicit Int64BigIntImpl(int64_t value)
        : m_value(value)
        , m_sign(value < 0)
    {
    }

    explicit Int64BigIntImpl(uint64_t value)
        : m_value(value)
    { }

    static constexpr unsigned numDigits = isRegister64Bit() ? 1 : 2;

    ALWAYS_INLINE bool isZero() { return !m_value; }
    ALWAYS_INLINE bool sign() { return m_sign; }
    ALWAYS_INLINE unsigned length()
    {
        return isZero() ? 0 : numDigits;
    }
    ALWAYS_INLINE JSBigInt::Digit digit(unsigned i)
    {
        ASSERT_UNUSED(i, i < length());
#if CPU(REGISTER64)
        if (sign())
            return static_cast<JSBigInt::Digit>(negate(m_value));
        return m_value;
#else
        if (sign()) {
            return static_cast<JSBigInt::Digit>(negate(m_value) >> (32 * i));
        }
        return static_cast<JSBigInt::Digit>(m_value >> (32 * i));
#endif
    }

private:
    friend struct JSBigInt::ImplResult;
    uint64_t m_value;
    bool m_sign { false };
};
```

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -1439,15 +1490,68 @@ JSBigInt::ComparisonResult JSBigInt::compare(JSBigInt* x, JSBigInt* y)
+    ASSERT(x.isBigInt() && y.isBigInt());
+#if USE(BIGINT32)
+    if (x.isBigInt32() && y.isBigInt32()) {
+        int32_t x1 = x.asBigInt32();
+        int32_t y1 = y.asBigInt32();
```
Can you make sure your other patch has tests for this bug?

hyjorc1 replied:
`USE_BIGINT32` is not enabled.

kmiller68 replied:
Oh, whoops! Never mind me, I thought it was.

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -473,6 +475,55 @@ class Int32BigIntImpl {
+        return m_value;
+#else
+        ASSERT(!i || i == 1);
+        ASSERT(sizeof(JSBigInt::Digit) == 32);
```
If you want to keep this assert it might as well be a `static_assert`

kmiller68 reviewed (approved):
r=me with recommendations.

kmiller68 commented on Source/JavaScriptCore/runtime/JSBigInt.cpp:
```diff
@@ -440,6 +440,10 @@ class HeapBigIntImpl {
 
+// For use in places where we could negate std::numeric_limits<T>::min and would like to avoid UB.
+template<std::integral T>
+ALWAYS_INLINE T negate(T v) { return ~static_cast<std::make_unsigned_t<T>>(v) + 1; }
```
Oh I would have put this in `WTF/MathExtras.h`

hyjorc1 replied:
Thanks. Yes, I missed that part.

hyjorc1 committed:
[JSC] Extend JSBigInt::compare and JSBigInt::compareToDouble for Int64 and Int32
https://bugs.webkit.org/show_bug.cgi?id=271073
rdar://124706111

Reviewed by Keith Miller.

This patch extends JSBigInt::compare and JSBigInt::compareToDouble
for Int64 and Int32.

* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::Int64BigIntImpl::Int64BigIntImpl):
(JSC::Int64BigIntImpl::isZero):
(JSC::Int64BigIntImpl::sign):
(JSC::Int64BigIntImpl::length):
(JSC::Int64BigIntImpl::digit):
(JSC::JSBigInt::compare):
(JSC::JSBigInt::compareToDouble):
* Source/JavaScriptCore/runtime/JSBigInt.h:

Canonical link: https://commits.webkit.org/276417@main
